check file with realPath is existing

drgnchan opened the pull request:
#26707 .Tomcat's `StandardContext.getRealPath(String path)` will return non null value when the file actually not exists.So add some code in `ServletContextResource` to check file is actually exists with jdk's `File` class.

drgnchan committed:
check file with realPath is existing

cuspymd reviewed (commented):
There is already an `exists()` method, do you think it is correct to check the existence of the file in the `isFile()` method?

drgnchan commented:
> There is already an `exists()` method, do you think it is correct to check the existence of the file in the `isFile()` method?

maybe it is a directory.

terminux committed:
Allow @Async qualifier to be declared with a placeholder or SpEL expression

Closes gh-27818

Commit:
Add CacheErrorHandler implementation that logs exceptions

See gh-27826

snicoll committed:
Provide an API to record various runtime hints

This commit provides an API to record the need for reflection,
resources, serialization, and proxies so that the runtime can be
optimized accordingly.

`RuntimeHints` provides an entry point to register the following:

* Reflection hints: individual elements of a type can be defined, as
well as a predefined categories (identified by the `MemberCategory`
enum). A method or constructor hint can refine whether the executable
should only be introspected or also invoked.
* Resource hints: patterns using includes/excludes identify the
resources to include at runtime. Resource bundles are also supported.
* Java Serialization hints: types that use java serialization can be
registered.
* Proxy hints: both interfaces-based (JDK) proxy and class-based proxy
can be defined.

This commit also introduces a `TypeReference` abstraction that permits
to record hints for types that are not available on the classpath, or
not compiled yet (generated code).

Closes gh-27829

snicoll committed:
Repackage Javapoet in org.springframework.javapoet

This commit repackages the Javapoet library into spring-core so that it
can be used by the AOT engine without requiring a specific version.

Closes gh-27828

rstoyanchev committed:
Minor refactoring in STOMP documentation

Closes gh-27906

rstoyanchev committed:
Consistent id for ReactorServerHttpRequest

Closes gh-27885

rstoyanchev committed:
Merge branch '5.3.x' into main

rstoyanchev committed:
Fix exception in Jetty10RequestUpgradeStrategy

Closes gh-27899

rstoyanchev committed:
Merge branch '5.3.x' into main

jhoeller committed:
Explicitly close InputStream after resolution in RequestPartMethodArgumentResolver

Closes gh-27773

jhoeller committed:
Skip "data:" lines without content

Closes gh-27923

jhoeller committed:
Upgrade to Hibernate ORM 5.4.33, Hibernate Validator 6.2.1, Mockito 4.2, AssertJ 3.22, HtmlUnit 2.56, XMLUnit 2.8.4

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle

jhoeller committed:
Upgrade to Hibernate ORM 5.6.3, Hibernate Validator 7.0.2, Jackson 2.13.1

sbrannen committed:
Infer UTF-8 encoding for JSON response with MockMvc's andDo(print(...))

Prior to this commit, the PrintingResultHandler in MockMvc -- typically
invoked via .andDo(print()) -- printed an `application/json` response
body using the default encoding (ISO-8859-1), which resulted in UTF-8
characters being garbled.

Since an `application/json` response is implicitly encoded using UTF-8,
the PrintingResultHandler now infers UTF-8 encoding for such response
bodies.

Closes gh-27926

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Polishing

snicoll committed:
Polishing

See gh-27926

snicoll committed:
Merge branch '5.3.x'

spring-builds committed:
Next development version (v5.3.16-SNAPSHOT)

bclozel committed:
Configure Spring Milestone repository for apiDiff task

Prior to this commit, the `apiDiff` custom Gradle task would only use
the configured repositories in the build to generated the API diff
report. This causes issues when the report has to be generated against a
previous milestone: the Spring Framework build only relies on Maven
Central and a specific `libs-spring-framework-build` repository for
building the project. In this case, the task cannot resolve the previous
milestone artifacts to generate the report.

This commit improves the `ApiDiffPlugin` to automatically add the Spring
Milestone repository to the root project configuration when the task is
executed.

Fixes gh-27928

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Improve comment parsing in DTD/XSD detection algorithm

Prior to this commit, XmlValidationModeDetector did not properly parse
all categories of comments (described below). When such categories of
comments were encountered XmlValidationModeDetector may have
incorrectly detected that an XML file used a DTD when it used an XSD,
or vice versa.

This commit revises the parsing algorithm in XmlValidationModeDetector
so that multi-line comments and multiple comments on a single line are
properly recognized.

Specifically, with this commit the following categories of comments are
now handled properly.

- Multiple comments on a single line
- Multi-line comment: beginning on one line and then ending on another
  line with an additional comment following on that same line.
- Multi-line comment: beginning at the end of XML content on one line
  and then spanning multiple lines.

Closes gh-27915

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Properly abort transferTo test for Undertow

Instead of simply returning prematurely and allowing the tests to be
marked as SUCCESS, this commit uses a failed assumption to abort the
the trasferTo tests for Undertow, resulting in the parameterized test
invocation properly being marked as ABORTED.

See gh-25310

sbrannen committed:
Merge branch '5.3.x'

JiangTChen committed:
Simplify debugging of Undertow transferTo issue

This commit introduces a dedicated (disabled) transferToWithUndertow()
test method to simplify debugging of transferTo issues with Undertow.

See gh-25310
Closes gh-27908

sbrannen committed:
Polish contribution

See gh-27908

mdeinum committed:
Fix regression in BeanPropertyRowMapper.underscoreName(String)

Commit 6316a35 introduced a regression for property names starting with
multiple uppercase letters (such as setEMail(...)).

This commit fixes that regression and includes an additional test to
cover this case.

See gh-27929
Closes gh-27941

jprinet committed:
Upgrade to Gradle Enterprise 3.8.1

Closes gh-27942

sbrannen committed:
Polishing

sbrannen committed:
Explicitly test BeanPropertyRowMapper.underscoreName(String)

See gh-27929

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Download gradle-enterprise-conventions plugin from "release" repo

See gh-27942

sbrannen committed:
Merge branch '5.3.x'

Commit:
Port GroovyDynamicElementReader to Java

Prior to this commit, GroovyDynamicElementReader was implemented in
Groovy, which required that developers have Groovy language support
installed and configured in their IDEs.

This commit ports GroovyDynamicElementReader from Groovy to Java,
making the compilation much simpler and allowing developers to open the
project in Eclipse or VSCode (or other IDEs without Groovy support)
without compiler errors.

This commit also converts related tests from Groovy to Java.

Closes gh-27945

sbrannen committed:
Ensure unresolvable placeholders can be ignored with @Value

Prior to this commit, if a PropertySourcesPlaceholderConfigurer bean
was configured with its ignoreUnresolvablePlaceholders flag set to
true, unresolvable placeholders in an @Value annotation were not
ignored, resulting in a BeanCreationException for the bean using @Value.

For example, given a property declared as `my.app.var = ${var}` without
a corresponding `var` property declared, an attempt to resolve
`@Value("${my.app.var}")` resulted in the following exception.

java.lang.IllegalArgumentException: Could not resolve placeholder 'var' in value "${var}"

This commit fixes this by modifying
PropertySourcesPlaceholderConfigurer's postProcessBeanFactory(...)
method so that a local PropertyResolver is created if the
ignoreUnresolvablePlaceholders flag is set to true. The local
PropertyResolver then enforces that flag, since the Environment in the
ApplicationContext is most likely not configured with
ignoreUnresolvablePlaceholders set to true.

Closes gh-27947

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Make PropertySourcesPlaceholderConfigurerTests pass on JDK 17

This change is required in Eclipse IDE but not for the Gradle build.

See gh-27947

sbrannen committed:
Apply "instanceof pattern matching" in PropertySourcesPlaceholderConfigurer

See gh-27947

jhoeller committed:
Avoid message listener recovery in case of persistence exceptions on commit

Closes gh-1807

jhoeller committed:
Polishing

jhoeller committed:
Upgrade to SLF4J 1.7.33, Netty 4.1.73, R2DBC Arabba-SR12

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle
#	spring-jms/src/main/java/org/springframework/jms/listener/AbstractPollingMessageListenerContainer.java

1chz committed:
Polish tests in spring-web

This PR polishes trivial things in tests in spring-web.

Closes gh-27958

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Polish contribution

See gh-27945

sbrannen committed:
Use AssertJ instead of Groovy Assertions

See gh-27945

sbrannen committed:
Polish GroovyApplicationContextTests

See gh-27945

sbrannen committed:
Overhaul GroovyBeanDefinitionReaderTests

This commit switches from CGLIB-based proxies to JDK dynamic proxies
for AOP and scoped-bean tests, in order to be able to run tests in the
Eclipse IDE without special Run Configuration setup for the Java module
system.

This commit also simplifies the loading of beans defined with the Groovy
DSL.

See gh-27945

sbrannen committed:
Remove Groovy from "Eclipse/STS Project Import Guide"

Groovy support is no longer required in the IDE due to gh-27945.

sbrannen committed:
Fix Javadoc links to JSR 305 annotations where feasible

Prior to this commit, we only (indirectly) configured external Javadoc
links for types in the javax.annotation package for JSR 250, since
those types are part of Java 8 SE. However, an external Javadoc link
for types from JSR 305 was not configured.

To address this issue, this commit now configures an external Javadoc
link for JSR 305 types. However, the external Javadoc link for JSR 305
must be configured last to ensure that types from JSR 250 (such as
@PostConstruct) are still supported. This is due to the fact that JSR
250 and JSR 305 both define types in javax.annotation, which results in
a split package, and the javadoc tool does not support split packages
across multiple external Javadoc sites. Instead, the Javadoc tool
iterates over all external links, and the first external site that
claims to support a given package (via the package-list file) wins.

This means:

- Javadoc for JSR 250 annotations is fully supported.
- Javadoc for JSR 305 annotations in the javax.annotation package will
  continue to result in a 404 (page not found) error.
- Javadoc for JSR 305 annotations in the javax.annotation.meta package
  is fully supported.

For Spring Framework 6.0, the Javadoc for JSR 250 types in
jakarta.annotation package is served from the JBoss Application Server
Javadoc site instead of from Oracle's Java SE documentation site, since
JSR 250 annotations are no longer included in Java SE.

Closes gh-27904

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Avoid use of `var` declarations

See gh-27945.

arey committed:
Fix CaffeineCacheManager configuration example in reference doc

See gh-27967

heowc committed:
MissingServletRequestPartException extends ServletRequestBindingException

Closes gh-27948

snicoll committed:
Merge pull request #27967 from arey

* pr/27967:
  Fix CaffeineCacheManager configuration example in reference doc

Closes gh-27967

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Polish ReflectionUtils[Tests]

sbrannen committed:
Filter methods in Object in ReflectionUtils.USER_DECLARED_METHODS

Prior to this commit, the USER_DECLARED_METHODS MethodFilter in
ReflectionUtils did not actually filter methods declared in
java.lang.Object as stated in its Javadoc.

Consequently, if ReflectionUtils.doWithMethods was invoked with
USER_DECLARED_METHODS and Object.class as the class to introspect, all
non-bridge non-synthetic methods declared in java.lang.Object were
passed to the supplied MethodCallback, which breaks the contract of
USER_DECLARED_METHODS.

In addition, if USER_DECLARED_METHODS was composed with a custom
MethodFilter using `USER_DECLARED_METHODS.and(<custom MethodFilter>)`,
that composed method filter allowed all non-bridge non-synthetic
methods declared in java.lang.Object to be passed to the supplied
MethodCallback, which also breaks the contract of
USER_DECLARED_METHODS. This behavior resulted in regressions in code
that had previously used USER_DECLARED_METHODS by itself and then began
using USER_DECLARED_METHODS in a composed filter. For example, since
commit c419ea7ba7 ReflectiveAspectJAdvisorFactory has incorrectly
processed methods in java.lang.Object as candidates for advice methods.

This commit fixes this bug and associated regressions by ensuring that
USER_DECLARED_METHODS actually filters methods declared in
java.lang.Object. In addition, ReflectionUtils.doWithMethods now aborts
its search algorithm early if invoked with the USER_DECLARED_METHODS
filter and Object.class as the class to introspect.

Closes gh-27970

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Polish ReflectionUtilsTests

sbrannen committed:
Enable ReflectionUtilsTests.findMethodWithVarArgs()

See gh-13286

sbrannen committed:
Merge branch '5.3.x'

jhoeller committed:
Check open status before close call (aligned with EntityManagerFactoryUtils)

Closes gh-27972

jhoeller committed:
Upgrade to Tomcat 9.0.58, Protobuf 3.19.3, Rome 1.18, H2 2.1.210, SLF4J 1.7.35, Mockito 4.3.1, HtmlUnit 2.57, XMLUnit 2.9, Checkstyle 9.2.1

jhoeller committed:
Upgrade to SmallRye Mutiny 1.3.1, Aalto 1.3.1, Woodstox 6.2.8

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle
#	spring-orm/src/main/java/org/springframework/orm/hibernate5/SessionFactoryUtils.java

jhoeller committed:
Upgrade to Hibernate ORM 5.6.5, Eclipse Yasson 2.0.4, JRuby 9.3.3

sbrannen committed:
Delete unused XML config in AOP namespace tests

See gh-22246

sbrannen committed:
Polish AOP namespace tests

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Apply "instanceof pattern matching" in additional core classes

snicoll committed:
Polish "Add CacheErrorHandler implementation that logs exceptions"

See gh-27826

snicoll committed:
Merge pull request #27826 from Drezir

* pr/27826:
  Polish "Add CacheErrorHandler implementation that logs exceptions"
  Add CacheErrorHandler implementation that logs exceptions

Closes gh-27826

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Polish AspectJ documentation

sbrannen committed:
Improve documentation for implementing AspectJ around advice

Closes gh-27980

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Polish GenericWebApplicationContext and AnnotationConfigWebApplicationContext

snicoll committed:
Upgrade Java versions in CI image

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Document how to register annotated classes in a GenericWebApplicationContext

Closes gh-27778

sbrannen committed:
Merge branch '5.3.x'

poutsma committed:
Support multiple boundary buffers in MultipartParser

In a small minority of cases, the multipart boundary can spread across
three incoming buffers.

Prior to this commit, MultipartParser.BodyState only supported two
buffers. If the boundary is spread across three buffers, the first
buffer of the three is sent as a whole, even though it contains the
first bytes of the boundary.

This commit fixes this bug, by enqueuing all prior buffers in a queue,
and emitting the ones that cannot contain boundary bytes.

Closes gh-27939

poutsma committed:
Merge branch '5.3.x'

wkwkhautbois committed:
Fix ServletUriComponentsBuilder examples in ref docs

Closes gh-27984

sbrannen committed:
Polish contribution

See gh-27984

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Improve Javadoc for DatabasePopulator

See gh-27008

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Stop applying Groovy plugin for the root Gradle project

The root project does not rely on the Groovy plugin.

See gh-27945

sbrannen committed:
Stop applying Groovy plugin for the root Gradle project

The root project does not rely on the Groovy plugin.

See gh-27945

sbrannen committed:
Reduce build time by reducing shutdown wait period for Jetty tests

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Remove dependency management for groovy-test

groovy-test is no longer used in the build.

See gh-27945

sbrannen committed:
Improve log message when searching for default executor for async processing

Closes gh-27983

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Merge branch '5.3.x'

snicoll committed:
Upgrade to spring javaformat 0.0.30

snicoll committed:
Merge branch '5.3.x'

rstoyanchev committed:
Ignore Content-Type that is invalid (not concrete)

Closes gh-27957

elgleidson committed:
Add support for strict JSON comparison in WebTestClient

Prior to this commit, WebTestClient only supported "lenient" comparison
of the expected JSON body.

This commit introduces an overloaded variant of `json()` in the
BodyContentSpec that accepts an additional boolean flag to specify
whether a "strict" comparison should be performed.

This new feature is analogous to the existing support in MockMvc.

Closes gh-27993

poutsma committed:
Consider current date in "1W" cron expressions

Prior to this commit, the QuartzCronField::weekdayNearestTo would elapse
until the next month before checking if the current day matched.

After this commit, the current day is checked before we elapse until
the next month.

Closes gh-27966

poutsma committed:
Merge branch '5.3.x'

sbrannen committed:
Apply "instanceof pattern matching" in spring-r2dbc

rstoyanchev committed:
Ensure all converters don't close InputStream

Closes gh-27969

sbrannen committed:
Apply "switch expressions" where applicable

sbrannen committed:
Sync ScriptUtilsUnitTests implementations

rstoyanchev committed:
Polishing tests

rstoyanchev committed:
Merge branch '5.3.x' into main

Remove use of Mockito spy proxies, which don't work on Java 17.

snicoll committed:
Upgrade Ubuntu version in CI image

snicoll committed:
Upgrade to spring javaformat 0.0.31

snicoll committed:
Merge branch '5.3.x'

bclozel committed:
Upgrade JDK 17 and JDK 18 versions in CI image

Commit:
Add formatting for SockJS GoAway frame

Prevents infinite loop for xhr-polling and xhr-streaming transports.

See gh-28000

snicoll committed:
Relocate runtime hints to aot package

See gh-27829

snicoll committed:
Polish

See gh-27829

sbrannen committed:
Use modern language features in tests

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Fix broken test missed in merge

sbrannen committed:
Use modern language features in tests

An1s9n committed:
Polish reference to ManagedBean annotation

See gh-28004

snicoll committed:
Merge pull request #28004 from An1s9n

* pr/28004:
  Polish reference to ManagedBean annotation

Closes gh-28004

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Ensure Spring AOP generates JDK dynamic proxies for lambdas

Prior to this commit, if AOP proxy generation was configured with
proxyTargetClass=true (which is the default behavior in recent versions
of Spring Boot), beans implemented as lambda expressions or method
references could not be proxied with CGLIB on Java 16 or higher without
specifying `--add-opens java.base/java.lang=ALL-UNNAMED`.

This commit addresses this shortcoming by ensuring that beans
implemented as lambda expressions or method references are always
proxied using a JDK dynamic proxy even if proxyTargetClass=true.

Closes gh-27971

sbrannen committed:
Merge branch '5.3.x'

jhoeller committed:
Upgrade to Groovy 4.0

Closes gh-27985

jhoeller committed:
Initial upgrade to R2DBC Borca

See gh-27976

jhoeller committed:
Consistently override (and prevent overriding for) aliases with bean definitions

Closes gh-27866

jhoeller committed:
Retrieve JDBC Connection via JdbcCoordinator (for compatibility with Hibernate 6)

Closes gh-28007

jhoeller committed:
Support for CGLIB BeanMap utility on JDK 17

Closes gh-27802

jhoeller committed:
Deprecate AsyncTaskExecutor.execute(Runnable task, long startTimeout)

Closes gh-27959

jhoeller committed:
Find interface method even for late-bound interface declaration in subclass

Closes gh-27995

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	spring-context-support/src/main/java/org/springframework/scheduling/commonj/WorkManagerTaskExecutor.java
#	spring-core/src/main/java/org/springframework/core/task/SimpleAsyncTaskExecutor.java
#	spring-expression/src/main/java/org/springframework/expression/spel/support/ReflectivePropertyAccessor.java
#	spring-tx/src/main/java/org/springframework/jca/work/SimpleTaskWorkManager.java
#	spring-tx/src/main/java/org/springframework/jca/work/WorkManagerTaskExecutor.java

jhoeller committed:
Update license header for https (nohttp rule)

See gh-27802

jhoeller committed:
Upgrade to Checkstyle 9.3, HtmlUnit 2.58, Apache HttpClient 5.1.3

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle

izeye committed:
Fix Javadoc since for AopProxyUtils.isLambda()

See gh-28010

Commit:
Apply "instanceof pattern matching" in AbstractRoutingDataSource

Closes gh-28011

snicoll committed:
Merge pull request #28010 from izeye

* pr/28010:
  Fix Javadoc since for AopProxyUtils.isLambda()

Closes gh-28010

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Polish contribution

See gh-28011

sbrannen committed:
Polishing

sbrannen committed:
Polish contribution

See gh-27993

sbrannen committed:
Polishing

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Polish JsonContentTests

Commit:
Use enhanced switch expressions where feasible

Closes gh-28014

sbrannen committed:
Ensure toString() for synthesized annotations is source code compatible

Since the introduction of synthesized annotation support in Spring
Framework 4.2 (a.k.a., merged annotations), the toString()
implementation attempted to align with the formatting used by the JDK
itself. However, Class annotation attributes were formatted using
Class#getName in Spring; whereas, the JDK used Class#toString up until
JDK 9.

In addition, JDK 9 introduced new formatting for toString() for
annotations, apparently intended to align with the syntax used in the
source code declaration of the annotation. However, JDK 9+ formats enum
annotation attributes using Enum#toString instead of Enum#name, which
can lead to issues if toString() is overridden in an enum.

This commit updates the formatting used for synthesized annotations by
ensuring that toString() generates a string that is compatible with the
syntax of the originating source code, going beyond the changes made in
JDK 9 by using Enum#name instead of Enum#toString.

Closes gh-28015

sbrannen committed:
Allow AutowiredAnnotationBeanPostProcessor to compile on JDK 11

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Remove unnecessary JDK 9+ check in MergedAnnotationsTests

sbrannen committed:
Apply "instanceof pattern matching" in SynthesizedMergedAnnotationInvocationHandler

sbrannen committed:
Upgrade to Gradle 7.4

Closes gh-28020

sbrannen committed:
Update Eclipse template to @since 5.3.16

snicoll committed:
Add convenience factory method for Managed[List|Set|Map]

Closes gh-28026

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Use canonical names for types in synthesized annotation toString

My proposal for the same change in the JDK is currently targeted for
JDK 19.

- https://bugs.openjdk.java.net/browse/JDK-8281462
- https://bugs.openjdk.java.net/browse/JDK-8281568
- https://github.com/openjdk/jdk/pull/7418

See gh-28015

snicoll committed:
Add core JavaPoet utilities

This commit adds utilities that facilitate code generation patterns
used by the AOT engine.

Closes gh-28028

snicoll committed:
Make BeanDefinitionValueResolver public

Closes gh-28029

snicoll committed:
Add code contribution infrastructure

This commit adds an API that lets individual components contribute code,
runtime hints, and protected access information. This ease the cases
where code need to be written in a privileged package if necessary and
let contributors provide hints for the code they generate.

Closes gh-28030

sbrannen committed:
Improve toString() for synthesized annotations

Although the initial report in gh-28015 only covered inconsistencies
for arrays and strings in the toString() implementations for
annotations between the JDK (after Java 9) and Spring, it has since
come to our attention that there was further room for improvement.

This commit therefore addresses the following in toString() output for
synthesized annotations.

- characters are now wrapped in single quotes.

- bytes are now properly formatted as "(byte) 0x##".

- long, float, and double values are now appended with "L", "f", and
  "d", respectively. The use of lowercase for "f" and "d" is solely to
  align with the choice made by the JDK team.

However, this commit does not address the following issues which we may
choose to address at a later point in time.

- non-ASCII, non-visible, and non-printable characters within a
  character or String literal are not escaped.

- formatting for float and double values does not take into account
  whether a value is not a number (NaN) or infinite.

Closes gh-28015

bclozel committed:
Upgrade to AspectJ 1.9.8

This commit also upgades the dedicated Gradle plugin for AspectJ build
integration.

Closes gh-27416

bclozel committed:
Upgrade to Kotlin 1.6.20-M1

This commit upgrades the Kotlin dependencies to 1.6.20-M1 and configures
the build to generate Java 17 bytecode for Kotlin classes.

Closes gh-27814

rstoyanchev committed:
Polishing contribution

See gh-27948

bclozel committed:
Upgrade Dokka to 1.6.10

sbrannen committed:
Merge branch '5.3.x'

# Conflicts:
#	spring-core/src/main/java/org/springframework/core/annotation/SynthesizedMergedAnnotationInvocationHandler.java
#	src/eclipse/org.eclipse.jdt.ui.prefs

sbrannen committed:
Simplify tests for synthesized annotation toString()

See gh-28015

sbrannen committed:
Merge branch '5.3.x'

# Conflicts:
#	spring-core/src/test/java/org/springframework/core/annotation/MergedAnnotationsTests.java

snicoll committed:
Start building against Reactor 2020.0.16 snapshots

See gh-28039

snicoll committed:
Merge branch '5.3.x'

sdeleuze committed:
Upgrade to Dokka 1.6.10

Closes gh-28040

sbrannen committed:
Short circuit if-conditions in AttributeMethods

sbrannen committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle

sbrannen committed:
Support recursive annotations in merged annotations

Although Java does not allow the definition of recursive annotations,
Kotlin does, and prior to this commit an attempt to synthesize a
merged annotation using the MergedAnnotation API resulted in a
StackOverflowError if there was a recursive cycle in the annotation
definitions.

This commit addresses this issue by tracking which annotations have
already been visited and short circuits the recursive algorithm if a
cycle is detected.

Closes gh-28012

vikeychen committed:
Fix CronExpression issue with DST

This commit fixes an issue with CronExpression fails to calculate next
execution on the day of daylight saving time.

Closes gh-28038

sbrannen committed:
Merge branch '5.3.x'

# Conflicts:
#	spring-core/src/test/java/org/springframework/core/annotation/AnnotationTypeMappingsTests.java

sdeleuze committed:
Make Kotlin functions accessible in CoroutinesUtils

In order to allow using private classes like in Java
for example.

Closes gh-23840

bclozel committed:
Upgrade CI pipeline

This commit upgrades the CI pipeline with the following:

* replace JDK16 with JDK17 as build variant
* upgrade all JDK versions
* replace docker-image with registry-image resource for CI image

bclozel committed:
Merge branch '5.3.x'

sdeleuze committed:
Merge branch '5.3.x'

bclozel committed:
Upgrade concourse-release-scripts in CI

This commit also reverts the change of resource type for the publication
of the CI image and fixes a bug in the CI image setup with available
JDKs.

bclozel committed:
Merge branch '5.3.x'

bclozel committed:
Configure Gradle JVM toolchain for Kotlin tasks

bclozel committed:
Remove toolchain configuration for Kotlin

Kotlin currently only supports JDK17 language level, which is our
minimum requirement so far.
This commit simplifies the toolchain setup until this is needed again.

See 303e363adfd

snicoll committed:
Add bean instance generator infrastructure

This commit provides the necessary infrastructure to let components
contribute statements that are used to fully instantiate a bean
instance.

To ease code generation, a dedicated infrastructure to register bean
definition is provided in the o.s.beans.factory.generator package.
BeanDefinitionRegistrar offers a builder style API that provides a way
to hide how injected elements are resolved at runtime and let
contributors provide code that may throw a checked exception.

BeanInstanceContributor is the interface that components can implement
to contribute to a bean instance setup. DefaultBeanInstanceGenerator
generates, for a particular bean definition, the necessary statements
to instantiate a bean.

Closes gh-28047

bclozel committed:
Make assertion message lazy in ServletRequestPathUtils

Closes gh-27946

bclozel committed:
Merge branch '5.3.x'

rstoyanchev committed:
Polishing contribution

Closes gh-28000

rstoyanchev committed:
Remove path variables from pathWithinMapping

Closes gh-27913

rstoyanchev committed:
Merge branch '5.3.x' into main

snicoll committed:
Introduce AOT variant of BeanPostProcessor

This commit introduces an optional interface that a BeanPostProcessor
can implement to opt-in for providing a generated code equivalent of
what it does with a regular runtime.

This commit has a first implementation of this interface with
AutowiredAnnotationBeanPostProcessor replacing its processing of
autowired annotations by generated code.

Closes gh-27921

snicoll committed:
Polish ProtectedAccess options

This commit improves how protected access analysis operates. Rather than
providing a static boolean, a function callback for the member to
analyse is used. This permits to change the decision whether reflection
can be used, or if the return type is assigned.

Both of those are already applicable, with InjectionGenerator relying
on reflection for private fields, and DefaultBeanInstanceGenerator
assigning the bean instance if additional contributors are present.

This commit also moves the logic of computing the options where the code
is actually generated.

See gh-28030

snicoll committed:
Upgrade to Reactor 2020.0.16

Closes gh-28039

snicoll committed:
Merge branch '5.3.x'

sbrannen committed:
Ensure fix for gh-28012 is actually tested

In 3ec612aaf8, I accidentally removed tests that verified support for
non-synthesizable merged annotations for recursive annotations in
Kotlin.

This commit reinstates those non-synthesizable tests while retaining
the synthesizable tests.

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Deprecate SocketUtils

SocketUtils was introduced in Spring Framework 4.0, primarily to assist
in writing integration tests which start an external server on an
available random port. However, these utilities make no guarantee about
the subsequent availability of a given port and are therefore
unreliable. Instead of using SocketUtils to find an available local
port for a server, it is recommended that users rely on a server's
ability to start on a random port that it selects or is assigned by the
operating system. To interact with that server, the user should query
the server for the port it is currently using.

SocketUtils is now deprecated in 5.3.16 and will be removed in 6.0.

Closes gh-28052

sbrannen committed:
Merge branch '5.3.x'

# Conflicts:
#	spring-web/src/test/java/org/springframework/remoting/caucho/CauchoRemotingTests.java
#	spring-webflux/src/test/java/org/springframework/web/reactive/function/client/WebClientIntegrationTests.java

sbrannen committed:
Polishing

sbrannen committed:
Increase max memory for Gradle to 2048m

sbrannen committed:
Stop using SocketUtils in WebClientIntegrationTests

See gh-28052

snicoll committed:
Polish

sbrannen committed:
Stop using SocketUtils in ActiveMQ STOMP broker tests

See gh-28052

sbrannen committed:
Remove deprecated SocketUtils

Closes gh-28052

philwebb committed:
Add mavenCentral() to pluginManagement repositories

Update the pluginManagement repositories used by Grade to include
`mavenCentral()` since `gradlePluginPortal()` has been suffering
from timeouts recently.

philwebb committed:
Add mavenCentral() to pluginManagement repositories

Update the pluginManagement repositories used by Grade to include
`mavenCentral()` since `gradlePluginPortal()` has been suffering
from timeouts recently.

philwebb committed:
Suppress deprecation warning in WebAsyncManagerTests

See gh-27959

poutsma committed:
Polish contribution

See gh-28038

poutsma committed:
Merge branch '5.3.x'

sbrannen committed:
Polish @Target declarations for stereotype annotations

sbrannen committed:
Improve documentation for uri(URI) method in WebTestClient

Prior to this commit, it was not clear that a configured base URI would
not be applied when invoking uri(URI).

This commit adds a note to the Javadoc to clarify that behavior.

Closes gh-28058

sbrannen committed:
Merge branch '5.3.x'

snicoll committed:
Merge branch '5.3.x'

jhoeller committed:
Upgrade to Netty 4.1.74, Jetty 9.4.45, Undertow 2.2.16, Hibernate Validator 6.2.2, Apache Johnzon 1.2.16, EclipseLink 2.7.10

jhoeller committed:
Polishing

jhoeller committed:
Merge branch '5.3.x'

# Conflicts:
#	build.gradle

jhoeller committed:
Upgrade to POI 5.2

spring-builds committed:
Next development version (v5.3.17-SNAPSHOT)

snicoll committed:
Downgrade to concourse-release-scripts 0.3.2

This commit reverts partially "0ab054c7b943d65bb9034d1d7987f556e9d54d05"
as 0.3.3 is breaking promition.

snicoll committed:
Merge branch '5.3.x'

poutsma committed:
Added .sdkmanrc file

This commit adds a .sdkmanrc file, so that we can automatically switch
to JDK 8 when building the 5.3. branch.

poutsma committed:
Merge branch '5.3.x'

poutsma committed:
Use JDK 17 on main branch in .sdkmanrc

This commit makes sure that JDK 17 is used on the main branch.

jhoeller committed:
Support for registering multiple init/destroy methods on AbstractBeanDefinition

Closes gh-28013

jhoeller committed:
Prevent @Bean method overloading by default (with enforceUniqueMethods flag)

Closes gh-22609

philwebb committed:
Polish BeanDefinitionRegistrar

sbrannen committed:
Recover from error during SpEL MIXED mode compilation

Prior to this commit, SpEL was able to recover from an error that
occurred while running a CompiledExpression; however, SpEL was not able
to recover from an error that occurred while compiling the expression
(such as a java.lang.VerifyError). The latter can occur when multiple
threads concurrently change types involved in the expression, such as
the concrete type of a custom variable registered via
EvaluationContext.setVariable(...), which can result in SpEL generating
invalid bytecode.

This commit addresses this issue by catching exceptions thrown while
compiling an expression and updating the `failedAttempts` and
`interpretedCount` counters accordingly. If an exception is caught
while operating in SpelCompilerMode.IMMEDIATE mode, the exception will
be propagated via a SpelEvaluationException with a new
SpelMessage.EXCEPTION_COMPILING_EXPRESSION error category.

Closes gh-28043

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Suppress deprecation warnings in tests in build

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Suppress deprecation warnings in tests in build

sbrannen committed:
Polish contribution

See gh-28014

sbrannen committed:
Polishing

sbrannen committed:
Polish contribution

See gh-27818

sbrannen committed:
Deprecate "enclosing classes" search strategy for MergedAnnotations

The TYPE_HIERARCHY_AND_ENCLOSING_CLASSES search strategy for
MergedAnnotations was originally introduced to support @Nested test
classes in JUnit Jupiter (see #23378).

However, while implementing #19930, we determined that the
TYPE_HIERARCHY_AND_ENCLOSING_CLASSES search strategy unfortunately
could not be used since it does not allow the user to control when to
recurse up the enclosing class hierarchy. For example, this search
strategy will automatically search on enclosing classes for static
nested classes as well as for inner classes, when the user probably
only wants one such category of "enclosing class" to be searched.
Consequently, TestContextAnnotationUtils was introduced in the Spring
TestContext Framework to address the shortcomings of the
TYPE_HIERARCHY_AND_ENCLOSING_CLASSES search strategy.

Since this search strategy is unlikely to be useful to general users,
the team has decided to deprecate this search strategy in Spring
Framework 5.3.x and remove it in 6.0.

Closes gh-28079

sbrannen committed:
Fix (@)since tag in SpelMessage

See gh-28043

sbrannen committed:
Merge branch '5.3.x'

sbrannen committed:
Remove deprecated "enclosing classes" search strategy for MergedAnnotations

Closes gh-28080

sbrannen committed:
Clean up warnings in test fixture

snicoll committed:
Polish naming

Co-authored-by: Andy Wilkinson <wilkinsona@vmware.com>

izeye committed:
Fix library name for JavaPoet in src/docs/dist/license.txt

This commit also updates Objenesis version in the file to align with the
version that Spring Framework is using.

See gh-28100

snicoll committed:
Merge pull request #28100 from izeye

* pr/28100:
  Fix library name for JavaPoet in src/docs/dist/license.txt

Closes gh-28100

snicoll committed:
Fix Objenesis version

See gh-28100

snicoll committed:
Merge branch '5.3.x'

poutsma committed:
Refactor roll forward in CronField

Before this commit, CronField.Type::rollForward added temporal units
to reach the higher order field. This caused issues with DST, where
the added amount of hours was either too small or too large.

This commit refactors the implementation so that it now adds one to the
higher order field, and reset the current field to the minimum value.

Closes gh-28095

poutsma committed:
Merge branch '5.3.x'

Commit:
Merge remote-tracking branch 'origin/master'

bclozel commented:
Closing this PR in favor of #26707 - your contribution has been merged in b71e686cbd, but this outdated PR got wrongly updated with lots of duplicate commits (a rebase and a force-push without any merge would have made things better).

Sorry it took us so long to process this PR, this fix is going to be shipped with the next 5.3.x maintenance release.
